### PR TITLE
upstream: load distribution in Total Panic mode

### LINKED ISSUE #4685: upstream: redesign per priority load calculation when all priority levels are in panic mode
Description:
Currently per priority traffic load distribution is calculated based on number of healthy hosts in each priority level. When there is zero healthy hosts in all priority levels, 100% of traffic goes to priority 0. 
This behavior should be changed for situation when all priority levels are in panic mode. It means that there is very low number of healthy hosts in each priority, possibly none. For this scenario (all priority levels are in panic mode) load distribution algorithm should use total number of hosts in each priority, not number of healthy hosts. For example if there are 3 priorities with 5 hosts each and the number of healthy hosts is 0 (P0), 1 (P1) and 1 (P2), the load will be 34% (33% plus rounding to 100%), 33% and 33% respectively.

### PR DESCRIPTION
**Pull request by cpakulski**
Description: Changed how load is calculated when all priority levels are in panic mode. Each priority level receives percentage of the traffic related to the number of hosts in that priority regardless of the health status of the hosts. This smooths out how traffic is shifted when hosts become unhealthy. See #4685 for design proposal and discussion.

Risk Level: Medium

Testing: Unit tests.

Docs Changes: Yes - updated section about panic threshold.

Release Notes: N/A
Fixes #4685

### CODE REVIEW

File: docs/root/intro/arch_overview/upstream/load_balancing/panic_threshold.rst [hunk 1]
```diff
@@ -79,8 +79,16 @@ priority.
 
 Panic mode can be disabled by setting the panic threshold to 0%.
 
-If all hosts become unhealthy normalized total health is 0%, and if the panic threshold is above 0%
-all traffic will be redirected to P=0.
+If the number of healthy hosts keeps decreasing, individual priority levels will start entering panic mode.
+Until at least one priority level is not in panic mode, load distribution is calculated as it is
```
Comment (snowp) [inline docs/root/intro/arch_overview/upstream/load_balancing/panic_threshold.rst]: I think flipping this sentence around would read a bit better: `Load distribution is calculated as describe above as long as there are priority levels not in panic mode`.

File: docs/root/intro/arch_overview/upstream/load_balancing/panic_threshold.rst [hunk 2]
```diff
@@ -79,8 +79,16 @@ priority.
 
 Panic mode can be disabled by setting the panic threshold to 0%.
 
-If all hosts become unhealthy normalized total health is 0%, and if the panic threshold is above 0%
-all traffic will be redirected to P=0.
+If the number of healthy hosts keeps decreasing, individual priority levels will start entering panic mode.
+Until at least one priority level is not in panic mode, load distribution is calculated as it is 
+described above. When all priority levels enter the panic mode, load calculation algorithm changes.
+In such state each priority level receives percentage of the traffic which represents the number
```
Comment (snowp) [inline docs/root/intro/arch_overview/upstream/load_balancing/panic_threshold.rst]: `In this case`

`receives traffic relative to the number of hosts in that priority level`

File: source/common/upstream/load_balancer_impl.cc [hunk 3]
```diff
@@ -236,22 +236,67 @@ void LoadBalancerBase::recalculatePerPriorityPanic() {
...
-  if (panic_threshold > 0 && normalized_total_availability == 0) {
-    // Everything is terrible. All load should be to P=0. Turn on panic mode.
-    ASSERT(per_priority_load_.healthy_priority_load_.get()[0] == 100);
-    per_priority_panic_[0] = true;
+  // This is corner case when panic is disabled and there is no hosts available.
+  // But the load must be distributed somewhere.
+  // This is just a placeholder, because no traffic will be directed to
+  // priority 0 anyways, because it is empty.
+  if (panic_threshold == 0 && normalized_total_availability == 0) {
+    per_priority_load_.healthy_priority_load_.get()[0] = 100;
```
Comment (snowp) [inline source/common/upstream/load_balancer_impl.cc]: How come this wasn't necessary previously?

File: source/common/upstream/load_balancer_impl.cc [hunk 4]
```diff
@@ -236,22 +236,67 @@ void LoadBalancerBase::recalculatePerPriorityPanic() {
...
   }
 
+  bool total_panic = true;
   for (size_t i = 0; i < per_priority_health_.get().size(); ++i) {
     // For each level check if it should run in panic mode. Never set panic mode if
     // normalized total health is 100%, even when individual priority level has very low # of
     // healthy hosts.
     const HostSet& priority_host_set = *priority_set_.hostSetsPerPriority()[i];
     per_priority_panic_[i] =
         (normalized_total_availability == 100 ? false : isGlobalPanic(priority_host_set));
```
Comment (snowp) [inline source/common/upstream/load_balancer_impl.cc]: does the function name `isGlobalPanic` still make sense? It seems a bit confusing to have both "total panic" and "global panic" - maybe we should rename the function?

File: source/common/upstream/load_balancer_impl.cc [hunk 5]
```diff
@@ -236,22 +236,67 @@ void LoadBalancerBase::recalculatePerPriorityPanic() {
...
+// are in panic mode. The load distribution is done NOT based on number
+// of healthy hosts in the priority, but based on number of hosts
+// in each priority regardless of its health.
+void LoadBalancerBase::recalculateLoadInTotalPanic() {
+  // first calculate total number of hosts across all priorities regardless
+  // whether they are healthy or not.
+  uint32_t total_hosts_num = 0;
+  for (size_t i = 0; i < per_priority_panic_.size(); i++) {
+    const HostSet& host_set = *priority_set_.hostSetsPerPriority()[i];
+    total_hosts_num += host_set.hosts().size() - host_set.excludedHosts().size();
```
Comment (snowp) [inline source/common/upstream/load_balancer_impl.cc]: I think `excludedHosts` are routed to in panic, so we probably want to include them here? I could be wrong, would be good to verify

File: source/common/upstream/load_balancer_impl.cc [hunk 6]
```diff
@@ -236,22 +236,67 @@ void LoadBalancerBase::recalculatePerPriorityPanic() {
...
+
+// recalculateLoadInTotalPanic method is called when all priority levels
+// are in panic mode. The load distribution is done NOT based on number
+// of healthy hosts in the priority, but based on number of hosts
+// in each priority regardless of its health.
+void LoadBalancerBase::recalculateLoadInTotalPanic() {
+  // first calculate total number of hosts across all priorities regardless
+  // whether they are healthy or not.
+  uint32_t total_hosts_num = 0;
+  for (size_t i = 0; i < per_priority_panic_.size(); i++) {
```
Comment (snowp) [inline source/common/upstream/load_balancer_impl.cc]: you can use `std::accumulate` here

File: source/common/upstream/load_balancer_impl.cc [hunk 7]
```diff
@@ -236,22 +236,67 @@ void LoadBalancerBase::recalculatePerPriorityPanic() {
...
 
-  // Panic mode is disabled only when panic_threshold is 0%.
-  if (panic_threshold > 0 && normalized_total_availability == 0) {
-    // Everything is terrible. All load should be to P=0. Turn on panic mode.
-    ASSERT(per_priority_load_.healthy_priority_load_.get()[0] == 100);
-    per_priority_panic_[0] = true;
+  // This is corner case when panic is disabled and there is no hosts available.
+  // But the load must be distributed somewhere.
+  // This is just a placeholder, because no traffic will be directed to
+  // priority 0 anyways, because it is empty.
```
Comment (snowp) [inline source/common/upstream/load_balancer_impl.cc]: technically its either empty or has no healthy hosts, right? would be good to be specific

File: source/common/upstream/load_balancer_impl.cc [hunk 8]
```diff
@@ -236,22 +236,67 @@ void LoadBalancerBase::recalculatePerPriorityPanic() {
...
+    recalculateLoadInTotalPanic();
+  }
+}
+
+// recalculateLoadInTotalPanic method is called when all priority levels
+// are in panic mode. The load distribution is done NOT based on number
+// of healthy hosts in the priority, but based on number of hosts
+// in each priority regardless of its health.
+void LoadBalancerBase::recalculateLoadInTotalPanic() {
+  // first calculate total number of hosts across all priorities regardless
```
Comment (snowp) [inline source/common/upstream/load_balancer_impl.cc]: nit: proper grammar in comments (capitalize `first`)

File: source/common/upstream/load_balancer_impl.cc [hunk 9]
```diff
@@ -236,22 +236,67 @@ void LoadBalancerBase::recalculatePerPriorityPanic() {
...
+  }
+  if (0 == total_hosts_num) {
+    // Backend is empty, but load must be distributed somewhere.
+    per_priority_load_.healthy_priority_load_.get()[0] = 100;
+    return;
+  }
+
+  // Now iterate through all priority levels and calculate how much
+  // load is supposed to go to each priority. In panic mode the calculation
+  // is based n ot on the number of healthy hosts but based on the number of
```
Comment (snowp) [inline source/common/upstream/load_balancer_impl.cc]: `n ot` -> `not`

File: source/common/upstream/load_balancer_impl.cc [hunk 10]
```diff
@@ -236,22 +236,67 @@ void LoadBalancerBase::recalculatePerPriorityPanic() {
...
+  // total hosts in the priority.
+  uint32_t total_load = 100;
+  for (size_t i = 0; i < per_priority_panic_.size(); i++) {
+    const HostSet& host_set = *priority_set_.hostSetsPerPriority()[i];
+    auto hosts_num = host_set.hosts().size() - host_set.excludedHosts().size();
```
Comment (snowp) [inline source/common/upstream/load_balancer_impl.cc]: nit: `const` where possbile (here and next line)

File: source/common/upstream/load_balancer_impl.cc [hunk 11]
```diff
@@ -236,22 +236,67 @@ void LoadBalancerBase::recalculatePerPriorityPanic() {
...
+
+    uint32_t priority_load = 100 * hosts_num / total_hosts_num;
+    per_priority_load_.healthy_priority_load_.get()[i] = priority_load;
+    per_priority_load_.degraded_priority_load_.get()[i] = 0;
+    total_load -= priority_load;
+  }
+  // add to priority zero whatever if left
```
Comment (snowp) [inline source/common/upstream/load_balancer_impl.cc]: nit: proper casing + period

File: test/common/upstream/load_balancer_impl_test.cc [hunk 12]
```diff
@@ -130,8 +130,18 @@ TEST_P(LoadBalancerBaseTest, PrioritySelection) {
   EXPECT_CALL(context, determinePriorityLoad(_, _)).WillRepeatedly(ReturnRef(priority_load));
   // With both the primary and failover hosts unhealthy, we should select an
   // unhealthy primary host.
-  EXPECT_EQ(100, lb_.percentageLoad(0));
-  EXPECT_EQ(0, lb_.percentageLoad(1));
+  // primary and failover are in panic mode. Load distribution is based
```
Comment (snowp) [inline test/common/upstream/load_balancer_impl_test.cc]: nit: `Primary`

File: test/common/upstream/load_balancer_impl_test.cc [hunk 13]
```diff
@@ -130,8 +130,18 @@ TEST_P(LoadBalancerBaseTest, PrioritySelection) {
   EXPECT_CALL(context, determinePriorityLoad(_, _)).WillRepeatedly(ReturnRef(priority_load));
   // With both the primary and failover hosts unhealthy, we should select an
```
Comment (snowp) [inline test/common/upstream/load_balancer_impl_test.cc]: this comment is out of date now

File: test/common/upstream/load_balancer_impl_test.cc [hunk 14]
```diff
@@ -130,8 +130,18 @@ TEST_P(LoadBalancerBaseTest, PrioritySelection) {
...
-  EXPECT_EQ(0, lb_.percentageLoad(1));
+  // primary and failover are in panic mode. Load distribution is based
+  // on the number of hosts regardless of their heath.
+  EXPECT_EQ(50, lb_.percentageLoad(0));
+  EXPECT_EQ(50, lb_.percentageLoad(1));
+  EXPECT_EQ(&host_set_, &lb_.chooseHostSet(&context).first);
+
+  // Modify number of hosts in failover, but leave them in unhealthy state
+  // primary and secondary are in panic mode, so load distribution is
+  // based on number of host regardless on their health.
```
Comment (snowp) [inline test/common/upstream/load_balancer_impl_test.cc]: `regardless of their health`

File: test/common/upstream/load_balancer_impl_test.cc [hunk 15]
```diff
@@ -130,8 +130,18 @@ TEST_P(LoadBalancerBaseTest, PrioritySelection) {
...
   // unhealthy primary host.
-  EXPECT_EQ(100, lb_.percentageLoad(0));
-  EXPECT_EQ(0, lb_.percentageLoad(1));
+  // primary and failover are in panic mode. Load distribution is based
+  // on the number of hosts regardless of their heath.
+  EXPECT_EQ(50, lb_.percentageLoad(0));
+  EXPECT_EQ(50, lb_.percentageLoad(1));
+  EXPECT_EQ(&host_set_, &lb_.chooseHostSet(&context).first);
+
+  // Modify number of hosts in failover, but leave them in unhealthy state
```
Comment (snowp) [inline test/common/upstream/load_balancer_impl_test.cc]: `in an/the unhealthy state`

File: test/common/upstream/load_balancer_impl_test.cc [hunk 16]
```diff
@@ -403,17 +415,24 @@ TEST_P(LoadBalancerBaseTest, GentleFailoverWithExtraLevels) {
   // All levels are completely down. 100% of traffic should go to P=0
```
Comment (snowp) [inline test/common/upstream/load_balancer_impl_test.cc]: update comment

File: test/common/upstream/load_balancer_impl_test.cc [hunk 17]
```diff
@@ -403,17 +415,24 @@ TEST_P(LoadBalancerBaseTest, GentleFailoverWithExtraLevels) {
...
-  ASSERT_THAT(getLoadPercentage(), ElementsAre(100, _, _));
+  updateHostSet(failover_host_set_, 3 /* num_hosts */, 0 /* num_healthy_hosts */);
+  updateHostSet(tertiary_host_set_, 2 /* num_hosts */, 0 /* num_healthy_hosts */);
+  ASSERT_THAT(getLoadPercentage(), ElementsAre(50, 30, 20));
   ASSERT_THAT(getDegradedLoadPercentage(), ElementsAre(0, 0, 0));
-  ASSERT_THAT(getPanic(), ElementsAre(true, _, _));
+  ASSERT_THAT(getPanic(), ElementsAre(true, true, true));
 
   // Rounding errors should be picked up by the first healthy priority.
+  // This situation is called TotalPanic, as all priority levels are in panic
```
Comment (snowp) [inline test/common/upstream/load_balancer_impl_test.cc]: its a bit odd that this explanation appears here when the above example is also in total panic, maybe move it up?

File: test/common/upstream/load_balancer_impl_test.cc [hunk 18]
```diff
@@ -427,13 +446,32 @@ TEST_P(LoadBalancerBaseTest, GentleFailoverWithExtraLevels) {
 
   // Rounding errors should be picked up by the first priority with degraded hosts when
   // there are no healthy priorities.
+  // Just disable panic threshold. Otherwise TotalPanic calculation
```
Comment (snowp) [inline test/common/upstream/load_balancer_impl_test.cc]: minor nit but I would rephrase this as `Disable panic threshold to prevent total panic from kicking in`. Alternatively, tweak the parameters of the test so that we're still able to test the behavior of rounding when everything is degraded without having to disable panic.

**Comment by cpakulski** [comment]
@snowp Thank you for code review. I am working on addressing your comments. There is one corner case which popped up during unit testing. This is when there is no heathy hosts across all priority levels. After my latest changes the algorithm just keeps running in total panic mode and distributes load to all priorities. However, there are some assumptions in the existing code that if there are no healthy hosts, chooseHost method should return nullptr. Interestingly, not all LBs have such assumptions. Ring LB has and its unit tests fail. Also aggregate cluster unit tests also assume that chooseHost returns nullptr. I think that TotalPanic mode should apply only when all levels are in panic mode and there is at least one healthy host in the cluster. I will change my code to work that way and will push code again.

**Comment by snowp** [comment]
That seems like an odd behavior, I would have expected panic mode to trigger even when there are no healthy hosts. Otherwise you might be saved by panic mode, only for your traffic to start failing once all hosts go unhealthy. @mattklein123 do you know the what the original intended behavior was around panic routing /w no healthy upstreams?

**Comment by mattklein123** [comment]
> That seems like an odd behavior, I would have expected panic mode to trigger even when there are no healthy hosts. Otherwise you might be saved by panic mode, only for your traffic to start failing once all hosts go unhealthy. @mattklein123 do you know the what the original intended behavior was around panic routing /w no healthy upstreams? 

Definitely a bug. Please fix!

**Comment by cpakulski** [comment]
Thanks for clarification! Will adjust the code accordingly.

**Comment by cpakulski** [comment]
wip: ping

File: docs/root/intro/arch_overview/upstream/load_balancing/panic_threshold.rst (hunk 1 above)
Comment (cpakulski) [reply docs/root/intro/arch_overview/upstream/load_balancing/panic_threshold.rst]: Done.

File: source/common/upstream/load_balancer_impl.cc (hunk 3 above)
Comment (cpakulski) [reply source/common/upstream/load_balancer_impl.cc]: I get segfault in unit tests for corner cases, like when all priorities are completely empty. Previously it was not a problem, but now the total number of hosts in used in division, so if it is zero, I get segfault (div by zero).

File: source/common/upstream/load_balancer_impl.cc (hunk 4 above)
Comment (cpakulski) [reply source/common/upstream/load_balancer_impl.cc]: Good point. It was confusing for me too, but I got used to it. Renamed to _isHostSetInPanic_.

File: source/common/upstream/load_balancer_impl.cc (hunk 5 above)
Comment (cpakulski) [reply source/common/upstream/load_balancer_impl.cc]: I have to verify it.

File: source/common/upstream/load_balancer_impl.cc (hunk 8 above)
Comment (cpakulski) [reply source/common/upstream/load_balancer_impl.cc]: Done.

File: source/common/upstream/load_balancer_impl.cc (hunk 9 above)
Comment (cpakulski) [reply source/common/upstream/load_balancer_impl.cc]: Done.

File: source/common/upstream/load_balancer_impl.cc (hunk 11 above)
Comment (cpakulski) [reply source/common/upstream/load_balancer_impl.cc]: Done.

File: test/common/upstream/load_balancer_impl_test.cc (hunk 12 above)
Comment (cpakulski) [reply test/common/upstream/load_balancer_impl_test.cc]: Done.

File: test/common/upstream/load_balancer_impl_test.cc (hunk 14 above)
Comment (cpakulski) [reply test/common/upstream/load_balancer_impl_test.cc]: Done.

File: test/common/upstream/load_balancer_impl_test.cc (hunk 15 above)
Comment (cpakulski) [reply test/common/upstream/load_balancer_impl_test.cc]: Done.

File: test/common/upstream/load_balancer_impl_test.cc (hunk 17 above)
Comment (cpakulski) [reply test/common/upstream/load_balancer_impl_test.cc]: Added the same explanation to all cases when TotalPanic is used.

File: test/common/upstream/load_balancer_impl_test.cc (hunk 18 above)
Comment (cpakulski) [reply test/common/upstream/load_balancer_impl_test.cc]: Done.

File: test/common/upstream/load_balancer_impl_test.cc (hunk 16 above)
Comment (cpakulski) [reply test/common/upstream/load_balancer_impl_test.cc]: Done. Comment updated with new logic.

File: test/common/upstream/load_balancer_impl_test.cc (hunk 13 above)
Comment (cpakulski) [reply test/common/upstream/load_balancer_impl_test.cc]: Done.

File: source/common/upstream/load_balancer_impl.cc (hunk 6 above)
Comment (cpakulski) [reply source/common/upstream/load_balancer_impl.cc]: Done. Good idea. Thanks!

File: source/common/upstream/load_balancer_impl.cc (hunk 10 above)
Comment (cpakulski) [reply source/common/upstream/load_balancer_impl.cc]: Done.

File: docs/root/intro/arch_overview/upstream/load_balancing/panic_threshold.rst (hunk 2 above)
Comment (cpakulski) [reply docs/root/intro/arch_overview/upstream/load_balancing/panic_threshold.rst]: Done.

**Comment by cpakulski** [comment]
@snowp I addressed all your code review comments accept https://github.com/envoyproxy/envoy/pull/9343#discussion_r357990021. I have to check the code to understand how a backend node ends up in `excludedHost` list and if we should route the traffic there while in panic mode.

**Comment by cpakulski** [comment]
@snowp I addressed the last issue https://github.com/envoyproxy/envoy/pull/9343#discussion_r357990021. `excludedHosts` array is filled in https://github.com/envoyproxy/envoy/blob/v1.12.2/source/common/upstream/upstream_impl.cc#L855-L872 and traffic is router to `excludedHosts` in TotalPanic state. The code is ready for another pass.

File: source/common/upstream/load_balancer_impl.cc [hunk 19]
```diff
@@ -238,22 +238,68 @@ void LoadBalancerBase::recalculatePerPriorityPanic() {
...
+}
+
+// recalculateLoadInTotalPanic method is called when all priority levels
+// are in panic mode. The load distribution is done NOT based on number
+// of healthy hosts in the priority, but based on number of hosts
+// in each priority regardless of its health.
+void LoadBalancerBase::recalculateLoadInTotalPanic() {
+  // First calculate total number of hosts across all priorities regardless
+  // whether they are healthy or not.
+  const uint32_t total_hosts_num = std::accumulate(
```
Comment (snowp) [inline source/common/upstream/load_balancer_impl.cc]: nit: i think i'd prefer `total_hosts_count`

File: source/common/upstream/load_balancer_impl.cc [hunk 20]
```diff
@@ -238,22 +238,68 @@ void LoadBalancerBase::recalculatePerPriorityPanic() {
...
+  for (size_t i = 0; i < per_priority_panic_.size(); i++) {
+    const HostSet& host_set = *priority_set_.hostSetsPerPriority()[i];
+    auto hosts_num = host_set.hosts().size() - host_set.excludedHosts().size();
+
+    uint32_t priority_load = 100 * hosts_num / total_hosts_num;
+    per_priority_load_.healthy_priority_load_.get()[i] = priority_load;
+    per_priority_load_.degraded_priority_load_.get()[i] = 0;
+    total_load -= priority_load;
+  }
+  // Add to priority zero whatever if left.
```
Comment (snowp) [inline source/common/upstream/load_balancer_impl.cc]: Do we need to worry about whether P0 has any hosts? Should we assign the rounding error to the first non-empty priority?

File: source/common/upstream/load_balancer_impl.cc [hunk 21]
```diff
@@ -238,22 +238,68 @@ void LoadBalancerBase::recalculatePerPriorityPanic() {
...
+  // Now iterate through all priority levels and calculate how much
+  // load is supposed to go to each priority. In panic mode the calculation
+  // is based not on the number of healthy hosts but based on the number of
+  // total hosts in the priority.
+  uint32_t total_load = 100;
+  for (size_t i = 0; i < per_priority_panic_.size(); i++) {
+    const HostSet& host_set = *priority_set_.hostSetsPerPriority()[i];
+    auto hosts_num = host_set.hosts().size() - host_set.excludedHosts().size();
+
+    uint32_t priority_load = 100 * hosts_num / total_hosts_num;
```
Comment (snowp) [inline source/common/upstream/load_balancer_impl.cc]: nit: const

File: source/common/upstream/load_balancer_impl.cc [hunk 22]
```diff
@@ -238,22 +238,68 @@ void LoadBalancerBase::recalculatePerPriorityPanic() {
...
+  }
+
+  // Now iterate through all priority levels and calculate how much
+  // load is supposed to go to each priority. In panic mode the calculation
+  // is based not on the number of healthy hosts but based on the number of
+  // total hosts in the priority.
+  uint32_t total_load = 100;
+  for (size_t i = 0; i < per_priority_panic_.size(); i++) {
+    const HostSet& host_set = *priority_set_.hostSetsPerPriority()[i];
+    auto hosts_num = host_set.hosts().size() - host_set.excludedHosts().size();
```
Comment (snowp) [inline source/common/upstream/load_balancer_impl.cc]: where did we land on whether to include excluded hosts? if they are used during panic routing i think it would make sense to not subtract them here?

File: source/common/upstream/load_balancer_impl.cc (hunk 20 above)
Comment (cpakulski) [reply source/common/upstream/load_balancer_impl.cc]: That would require configuration with empty P0 and not empty P>0. To be on the safe side I added loop to find the first not empty priority. Thanks for suggestion.

File: source/common/upstream/load_balancer_impl.cc (hunk 22 above)
Comment (cpakulski) [reply source/common/upstream/load_balancer_impl.cc]: Done.

File: source/common/upstream/load_balancer_impl.cc (hunk 21 above)
Comment (cpakulski) [reply source/common/upstream/load_balancer_impl.cc]: Done.

File: source/common/upstream/load_balancer_impl.cc (hunk 22 above)
Comment (cpakulski) [reply source/common/upstream/load_balancer_impl.cc]: I found out that excluded hosts are those for which healthcheck is not resolved yet. I understand that this was added to avoid entering panic mode when bunch of new hosts is added for which health is not known yet. Whether to route to them in panic mode? I guess the answer is yes and that is how it is implemented now.

The existence of `host_set.excludedHosts().size();` here is a bug, because I removed it from calculation of the total load in my previous commit but I did not remove it here when hosts are calculated for each priority. I was perplexed why none of tests caught it. So I added a new test to include excludedHosts case, but the behavior was not what I expected. It happened that there was a bug in test setup routine. The following line is wrong: https://github.com/envoyproxy/envoy/blob/v1.12.2/test/common/upstream/load_balancer_impl_test.cc#L87. It was probably copy-pasted.

File: source/common/upstream/load_balancer_impl.cc (hunk 19 above)
Comment (cpakulski) [reply source/common/upstream/load_balancer_impl.cc]: Done.

File: source/common/upstream/load_balancer_impl.cc [hunk 23]
```diff
@@ -238,21 +238,74 @@ void LoadBalancerBase::recalculatePerPriorityPanic() {
...
+  }
+
+  // Now iterate through all priority levels and calculate how much
+  // load is supposed to go to each priority. In panic mode the calculation
+  // is based not on the number of healthy hosts but based on the number of
+  // total hosts in the priority.
+  uint32_t total_load = 100;
+  for (size_t i = 0; i < per_priority_panic_.size(); i++) {
+    const HostSet& host_set = *priority_set_.hostSetsPerPriority()[i];
+    const auto hosts_num = host_set.hosts().size();
```
Comment (snowp) [inline source/common/upstream/load_balancer_impl.cc]: nit: since this isnt reused i would just inline it

File: source/common/upstream/load_balancer_impl.cc [hunk 24]
```diff
@@ -238,21 +238,74 @@ void LoadBalancerBase::recalculatePerPriorityPanic() {
...
+
+    const uint32_t priority_load = 100 * hosts_num / total_hosts_count;
+    per_priority_load_.healthy_priority_load_.get()[i] = priority_load;
+    per_priority_load_.degraded_priority_load_.get()[i] = 0;
+    total_load -= priority_load;
+  }
+
+  // Find the first priority which is not empty and add remaining load.
+  for (size_t i = 0; i < per_priority_panic_.size(); i++) {
```
Comment (snowp) [inline source/common/upstream/load_balancer_impl.cc]: in other places we do this kind of computation we determine what priority to allocate the remainder to during the loop where we distribute load, see https://github.com/envoyproxy/envoy/blob/master/source/common/upstream/load_balancer_impl.cc#L205

i'd probably just keep a `size_t first_nonempty` variable that gets set to the index of the first priority with  hosts in it and then use that to allocate the remainder.

File: test/common/upstream/load_balancer_impl_test.cc [hunk 25]
```diff
@@ -402,21 +426,34 @@ TEST_P(LoadBalancerBaseTest, GentleFailoverWithExtraLevels) {
...
-  ASSERT_THAT(getLoadPercentage(), ElementsAre(100, _, _));
+  updateHostSet(failover_host_set_, 3 /* num_hosts */, 0 /* num_healthy_hosts */);
+  updateHostSet(tertiary_host_set_, 2 /* num_hosts */, 0 /* num_healthy_hosts */);
+  ASSERT_THAT(getLoadPercentage(), ElementsAre(50, 30, 20));
   ASSERT_THAT(getDegradedLoadPercentage(), ElementsAre(0, 0, 0));
-  ASSERT_THAT(getPanic(), ElementsAre(true, _, _));
+  ASSERT_THAT(getPanic(), ElementsAre(true, true, true));
 
-  // Rounding errors should be picked up by the first healthy priority.
+  // Rounding errors should be picked up by the first priority.
```
Comment (snowp) [inline test/common/upstream/load_balancer_impl_test.cc]: Can you add a test for when the priority that gets given the remainder is not the first priority?

File: test/extensions/clusters/aggregate/cluster_test.cc [hunk 26]
```diff
@@ -217,7 +217,16 @@ TEST_F(AggregateClusterTest, AllHostAreUnhealthyTest) {
   EXPECT_CALL(primary_load_balancer_, chooseHost(_)).WillRepeatedly(Return(host));
   EXPECT_CALL(secondary_load_balancer_, chooseHost(_)).WillRepeatedly(Return(nullptr));
 
-  for (int i = 0; i < 100; ++i) {
```
Comment (snowp) [inline test/extensions/clusters/aggregate/cluster_test.cc]: mind including a comment here about whats going on? its not super obvious from reading the code

File: source/common/upstream/load_balancer_impl.cc (hunk 23 above)
Comment (cpakulski) [reply source/common/upstream/load_balancer_impl.cc]: I am reusing it now for the logic related to rounding errors.

File: test/common/upstream/load_balancer_impl_test.cc (hunk 25 above)
Comment (cpakulski) [reply test/common/upstream/load_balancer_impl_test.cc]: Done.

File: source/common/upstream/load_balancer_impl.cc (hunk 24 above)
Comment (cpakulski) [reply source/common/upstream/load_balancer_impl.cc]: Done.

File: test/extensions/clusters/aggregate/cluster_test.cc (hunk 26 above)
Comment (cpakulski) [reply test/extensions/clusters/aggregate/cluster_test.cc]: Comments were added.

File: source/common/upstream/load_balancer_impl.cc (hunk 5 above)
Comment (cpakulski) [reply source/common/upstream/load_balancer_impl.cc]: Done. excludedHosts are not used when calculating load distribution in TotalPanic.

File: docs/root/intro/version_history.rst [hunk 27]
```diff
@@ -60,6 +60,7 @@ Version history
 * tracing: added initial support for AWS X-Ray (local sampling rules only) :ref:`X-Ray Tracing <envoy_api_msg_config.trace.v2alpha.XRayConfig>`.
 * tracing: added tags for gRPC request path, authority, content-type and timeout.
 * udp: added initial support for :ref:`UDP proxy <config_udp_listener_filters_udp_proxy>`
+* upstream: changed load distribution algorithm when all priorities enter :ref:`panic mode<arch_overview_load_balancing_panic_threshold>`.
```
Comment (alyssawilk) [inline docs/root/intro/version_history.rst]: This will need to be moved up to the 1.4 section since we just cut the quarterly release :-)

File: source/common/upstream/load_balancer_impl.cc (hunk 7 above)
Comment (alyssawilk) [reply source/common/upstream/load_balancer_impl.cc]: Yeah, this comment doesn't seem quite right.  P[0] may have hosts (it's not necessarily empty).  In theory it would just fail to accept queries right?

Comment (alyssawilk) [reply source/common/upstream/load_balancer_impl.cc]: Also I think there's a comment above recalculatePerPriorityState which I think may need updating.

File: source/common/upstream/load_balancer_impl.h [hunk 28]
```diff
@@ -58,7 +58,8 @@ class LoadBalancerBase : public LoadBalancer {
    * majority of hosts are unhealthy we'll be likely in a panic mode. In this case we'll route
    * requests to hosts regardless of whether they are healthy or not.
    */
-  bool isGlobalPanic(const HostSet& host_set);
+  bool isHostSetInPanic(const HostSet& host_set);
+  void recalculateLoadInTotalPanic();
```
Comment (alyssawilk) [inline source/common/upstream/load_balancer_impl.h]: Comment please

Comment (cpakulski) [reply source/common/upstream/load_balancer_impl.h]: Done.

File: docs/root/intro/version_history.rst (hunk 27 above)
Comment (cpakulski) [reply docs/root/intro/version_history.rst]: Done.

File: source/common/upstream/load_balancer_impl.cc (hunk 7 above)
Comment (cpakulski) [reply source/common/upstream/load_balancer_impl.cc]: Updated the comment above recalculatePerPriorityState. Added explanation why 100% of load is assigned to P=0 when there are no healthy hosts and panic is disabled. Other routines expect that the sum of all percentages adds to 100%. If this is not the case there are assertions which fail. When panic is disabled, the load balancers will look for healthy hosts and because there is none, error will be returned and nothing will be routed to backend cluster.

File: docs/root/intro/version_history.rst [hunk 29]
```diff
@@ -66,6 +67,7 @@ Version history
 * tracing: added initial support for AWS X-Ray (local sampling rules only) :ref:`X-Ray Tracing <envoy_api_msg_config.trace.v2alpha.XRayConfig>`.
 * tracing: added tags for gRPC request path, authority, content-type and timeout.
 * udp: added initial support for :ref:`UDP proxy <config_udp_listener_filters_udp_proxy>`
+* fault: fixed an issue where the http fault filter would repeatedly check the percentage of abort/delay when the `x-envoy-downstream-service-cluster` header was included in the request to ensure that the actual percentage of abort/delay matches the configuration of the filter.
```
Comment (snowp) [inline docs/root/intro/version_history.rst]: bad merge?

Comment (cpakulski) [reply docs/root/intro/version_history.rst]: Just corrected it and pushed again. Not sure how it happened. I was getting merge conflict for that file and probably hit wrong keys in vim. Thanks for spotting it.
